checks/permissions.go

```diff
@@ -24,149 +24,251 @@ import (
 
 	// Map type.
 	case map[interface{}]interface{}:
-		if res, err := validateMapPermissions(val, path, dl); err != nil {
-			return false, err
-		} else if !res {
-			return false, nil
+		if err := validateMapPermissions(val, path, dl, pdata); err != nil {
+			return err
 		}
 
 	// Invalid type.
 	default:
 		//nolint
-		return false, sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
+		return sce.Create(sce.ErrScorecardInternal, errInvalidGitHubWorkflowFile.Error())
 	}
 
-	return true, nil
+	return nil
+}
+
+func isPermissionOfInterest(name string) bool {
+	return strings.EqualFold(name, "statuses") ||
+		strings.EqualFold(name, "checks") ||
+		strings.EqualFold(name, "security-events") ||
+		strings.EqualFold(name, "deployments") ||
+		strings.EqualFold(name, "contents") ||
+		strings.EqualFold(name, "packages") ||
+		strings.EqualFold(name, "options")
+}
+
+// Calculate the score.
+func calculateScore(result permissionCbData) int {
+	// See list https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/.
+	// Note: there are legitimate reasons to use some of the permissions like checks, deployments, etc.
+	// in CI/CD systems https://docs.travis-ci.com/user/github-oauth-scopes/.
+	if _, ok := result.permissions["all"]; ok {
+		return checker.MinResultScore
+	}
+
+	score := float32(checker.MaxResultScore)
+	// status: https://docs.github.com/en/rest/reference/repos#statuses.
+	// May allow an attacker to change the result of pre-submit and get a PR merged.
+	// Low risk: -0.5.
+	if _, ok := result.permissions["statuses"]; ok {
+		score -= 0.5
+	}
+
+	// checks.
+	// May allow an attacker to edit checks to remove pre-submit and introduce a bug.
+	// Low risk: -1.
+	if _, ok := result.permissions["checks"]; ok {
+		score -= 0.5
+	}
+
+	// secEvents.
+	// May allow attacker to read vuln reports before patch available.
+	// Low risk: -1
+	if _, ok := result.permissions["security-events"]; ok {
+		score--
+	}
+
+	// deployments: https://docs.github.com/en/rest/reference/repos#deployments.
+	// May allow attacker to charge repo owner by triggering VM runs,
+	// and tiny chance an attacker can trigger a remote
+	// service with code they own if server accepts code/location var unsanitized.
+	// Low risk: -1
+	if _, ok := result.permissions["deployments"]; ok {
+		score--
+	}
+
+	// contents.
+	// Allows attacker to commit unreviewed code.
+	// High risk: -10
+	if _, ok := result.permissions["contents"]; ok {
+		score -= checker.MaxResultScore
+	}
+
+	// packages.
+	// Allows attacker to publish packages.
+	// High risk: -10
+	if _, ok := result.permissions["packages"]; ok {
+		score -= checker.MaxResultScore
+	}
+
+	// actions.
+	// May allow an attacker to steal GitHub secrets by adding a malicious workflow/action.
+	// High risk: -10
+	if _, ok := result.permissions["actions"]; ok {
+		score -= checker.MaxResultScore
+	}
+
+	if score < 0 {
```

Comment:
nit: use `checker.MinResultScore` in place of both 0s here?

Reply:
thanks